controllers/metal3remediation_controller.go

```diff
@@ -0,0 +1,222 @@
+	}
+	err = r.Get(ctx, key, &metal3Machine)
+	if err != nil {
+		remediationLog.Error(err, "metal3machine not found")
+		return ctrl.Result{}, errors.Wrapf(err, "metal3machine not found")
+	}
+
+	remediationLog = remediationLog.WithValues("metal3machine", metal3Machine.Name)
+
+	// Create a helper for managing the remediation object.
+	remediationMgr, err := r.ManagerFactory.NewRemediationManager(metal3Remediation, &metal3Machine, capiMachine, remediationLog)
+	if err != nil {
+		remediationLog.Error(err, "failed to create helper for managing the metal3remediation")
+		return ctrl.Result{}, errors.Wrapf(err, "failed to create helper for managing the metal3remediation")
+	}
+
+	// Handle deleted remediation
+	if !metal3Remediation.ObjectMeta.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(ctx, remediationMgr)
+	}
+
+	// Handle non-deleted remediation
+	return r.reconcileNormal(ctx, remediationMgr)
+}
+
+func (r *Metal3RemediationReconciler) reconcileNormal(ctx context.Context,
+	remediationMgr baremetal.RemediationManagerInterface,
+) (ctrl.Result, error) {
+
+	// If host is gone, exit early
+	host, _, err := remediationMgr.GetUnhealthyHost(ctx)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "unable to find a host for unhealthy machine")
+	}
+
+	// If user has set bmh.Spec.Online to false
+	// do not try to remediate the host
+	if !remediationMgr.OnlineStatus(host) {
+		r.Log.Info("Host is powered off, unable to remediate")
+		remediationMgr.SetRemediationPhase(capm3.PhaseFailed)
+		return ctrl.Result{}, nil
+	}
+
+	remediationType := remediationMgr.GetRemediationType()
+
+	if remediationType != capm3.RebootRemediationStrategy {
+		r.Log.Info("unsupported remediation strategy")
+		return ctrl.Result{}, nil
+	}
+
+	if remediationType == capm3.RebootRemediationStrategy {
+		// If no phase set, default to running
+		if remediationMgr.GetRemediationPhase() == "" {
+			remediationMgr.SetRemediationPhase(capm3.PhaseRunning)
+		}
+
+		switch remediationMgr.GetRemediationPhase() {
+		case capm3.PhaseRunning:
+			// host is not rebooted yet
+			if remediationMgr.GetLastRemediatedTime() == nil {
+				r.Log.Info("Rebooting the host")
+				err := remediationMgr.SetRebootAnnotation(ctx)
+				if err != nil {
+					return ctrl.Result{}, errors.Wrap(err, "error setting reboot annotation")
+				}
+				now := metav1.Now()
+				remediationMgr.SetLastRemediationTime(&now)
+				remediationMgr.IncreaseRetryCount()
+			}
+
+			if remediationMgr.RetryLimitIsSet() && !remediationMgr.HasReachRetryLimit() {
+				okToRemediate, nextRemediation := remediationMgr.TimeToRemediate(remediationMgr.GetTimeout().Duration)
+
+				if okToRemediate {
+					err := remediationMgr.SetRebootAnnotation(ctx)
+					if err != nil {
+						return ctrl.Result{}, errors.Wrapf(err, "error setting reboot annotation")
+					}
+					now := metav1.Now()
+					remediationMgr.SetLastRemediationTime(&now)
+					remediationMgr.IncreaseRetryCount()
```

Comment:
looks like this code block appears twice so it might be worth to extract it into a function